s/strnum.ss

```diff
@@ -204,13 +204,30 @@ an exception.
 
 (define noround (lambda (x) x))
 (define rounder
-  (lambda (p)
+  (lambda (p ex)
     (if (zero? p)
         (lambda (n) 0)
         (lambda (n)
           (let ([a (numerator n)]
                 [b (denominator n)])
-	    (let ([d (- (bitwise-length a) (bitwise-length b))])
+	    (let* ([a-bits (bitwise-length a)]
+                   [b-bits (bitwise-length b)]
+                   [d (- a-bits b-bits)]
+                   ;; If `p` is large, we might run out of memory by
+                   ;; shifting by it directly, but in some cases, the right
+                   ;; result should fit into memory no matter how big `p` is.
+                   [p (cond
+                        [(not (eq? ex 'e))
+                         ;; end result will have at most 53 bits, anyway, so
+                         ;; bound p; we don't need a tight bound, and adding 2
+                         ;; extra bits over `double` precision to make sure
+                         ;; rounding will be right
+                         (min p (+ (max a-bits b-bits) 53 2))]
```

Comment:
We are in the case where a binary fraction `x` is first rounded to `p` significant bits and then to 53 significant bits (round to even in case of a tie). This is, in general, not the same as rounding directly to 53 significant bits. The case where it may differ looks like `X.1Y`. Here, `X` represents 53 significant binary digits, the dot is the point of rounding and `Y` are all further binary digits. The rounding of `X.1Y` to 53 binary digits depends on whether `Y` vanishes or not. If we first round `X.1Y` to `p` significant digits, it may happen that a nonzero `Y` becomes zero after rounding. Because of this, we must make sure that if we truncate `p`, the answer to the question of whether `Y` becomes zero or not won't change.

Every binary fraction is a transient (non-repeating digits, possibly trivial) followed by a repetend (repeating digits, possibly just zeros). We are allowed to truncate `p` when `p` points into the repetend and remains there after truncation (because this won't change whether `Y` becomes zero after rounding).  So, for this, we have to estimate the length of the transient. This has at most `(+ a-bits b-bits)` bits. Together with the 53 bits for `X` and `2` bits for a safety measure to exclude possible edge cases, the formula should therefore be `(min p (+ a-bits b-bits 53 2))`.